File: Week3 Assignment/remove_loop.c
//Remove loop

#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

void insertAtEnd(struct Node** head, int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->next = NULL;

    if (*head == NULL) {
        *head = newNode;
        return;
    }

    struct Node* temp = *head;
    while (temp->next != NULL)
        temp = temp->next;
    temp->next = newNode;
}

void removeLoop(struct Node* loop_node, struct Node* head) {
    struct Node* ptr1 = head;
    struct Node* ptr2;

    // Find the start of loop
    while (1) {
        ptr2 = loop_node;
        while (ptr2->next != loop_node && ptr2->next != ptr1) {
            ptr2 = ptr2->next;
        }
        if (ptr2->next == ptr1)
            break;
        ptr1 = ptr1->next;
    }

    // Break the loop
    ptr2->next = NULL;
}

int detectAndRemoveLoop(struct Node* head) {
    struct Node *slow = head, *fast = head;

    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;

        // Loop detected
        if (slow == fast) {
            removeLoop(slow, head);
            return 1;
        }
    }
    return 0;
}

void printList(struct Node* head) {
    while (head) {
        printf("%d -> ", head->data);
        head = head->next;
    }
    printf("NULL\n");
}

int main() {
    struct Node* head = NULL;

    insertAtEnd(&head, 10);
    insertAtEnd(&head, 20);
    insertAtEnd(&head, 30);
    insertAtEnd(&head, 40);

    // Create a loop for testing: last node -> second node
    head->next->next->next->next = head->next;

    if (detectAndRemoveLoop(head))
        printf("Loop detected and removed\n");
    else
        printf("No loop found\n");

    printList(head);

    return 0;
}

File: Week3 Assignment/remove_loop.py
#Remove loop

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def insert_at_end(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            return
        temp = self.head
        while temp.next:
            temp = temp.next
        temp.next = new_node

    def detect_and_remove_loop(self):
        slow = self.head
        fast = self.head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

            # Loop detected
            if slow == fast:
                self.remove_loop(slow)
                return True
        return False

    def remove_loop(self, loop_node):
        ptr1 = self.head

        while True:
            ptr2 = loop_node
            while ptr2.next != loop_node and ptr2.next != ptr1:
                ptr2 = ptr2.next
            if ptr2.next == ptr1:
                break
            ptr1 = ptr1.next

        # Break the loop
        ptr2.next = None

    def print_list(self):
        temp = self.head
        while temp:
            print(temp.data, end=" -> ")
            temp = temp.next
        print("None")

# Example usage
ll = LinkedList()
ll.insert_at_end(10)
ll.insert_at_end(20)
ll.insert_at_end(30)
ll.insert_at_end(40)

# Create loop for testing: last node -> second node
ll.head.next.next.next.next = ll.head.next

if ll.detect_and_remove_loop():
    print("Loop detected and removed")
else:
    print("No loop found")

ll.print_list()
